### ISSUE TITLE
ansible-galaxy login -c CERT error

### ISSUE THREAD

**Issue by chrismeyersfsu** [ansible/ansible#59734]
##### SUMMARY
`ansible-galaxy login --ignore-certs` is not respected. Further, the traceback for the error when a cert fails isn't clear.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
ansible-galaxy login https://github.com/ansible/ansible/blob/devel/lib/ansible/galaxy/login.py#L91

##### ANSIBLE VERSION

```

```

##### CONFIGURATION

```
ansible 2.9.0.dev0
  config file = None
  configured module search path = [u'/home/meyers/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/meyers/ansible/virtualenv/ansible-dev2/local/lib/python2.7/site-packages/ansible
  executable location = /home/meyers/ansible/virtualenv/ansible-dev2/bin/ansible
  python version = 2.7.16 (default, Apr  6 2019, 01:42:57) [GCC 8.3.0]
```

##### OS / ENVIRONMENT

```
cat /etc/lsb-release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=19.04
DISTRIB_CODENAME=disco
DISTRIB_DESCRIPTION="Ubuntu 19.04"
```

##### STEPS TO REPRODUCE

```
ansible-galaxy login -c
```
You can enter junk for the username/password and the error will trigger.

##### EXPECTED RESULTS
```
ERROR! Bad credentials
```

##### ACTUAL RESULTS

```
kjfldjansible-galaxy 2.9.0.dev0
  config file = None
  configured module search path = [u'/home/meyers/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/meyers/ansible/virtualenv/ansible-dev2/local/lib/python2.7/site-packages/ansible
  executable location = /home/meyers/ansible/virtualenv/ansible-dev2/bin/ansible-galaxy
  python version = 2.7.16 (default, Apr  6 2019, 01:42:57) [GCC 8.3.0]
No config file found; using defaults
Opened /home/meyers/.ansible_galaxy


We need your GitHub login to identify you.
This information will not be sent to Galaxy, only to api.github.com.
The password will not be displayed.

Use --github-token if you do not want to enter your password.

GitHub Username: fldlfd
Password for dkjfldjfldlfd:
ca path None
Force False
ERROR! Unexpected Exception, this is probably a bug: Incorrect padding
the full traceback was:

Traceback (most recent call last):
  File "/home/meyers/ansible/virtualenv/ansible-dev2/bin/ansible-galaxy", line 111, in <module>
    exit_code = cli.run()
  File "/home/meyers/ansible/virtualenv/ansible-dev2/local/lib/python2.7/site-packages/ansible/cli/galaxy.py", line 269, in run
    context.CLIARGS['func']()
  File "/home/meyers/ansible/virtualenv/ansible-dev2/local/lib/python2.7/site-packages/ansible/cli/galaxy.py", line 888, in execute_login
    github_token = login.create_github_token()
  File "/home/meyers/ansible/virtualenv/ansible-dev2/local/lib/python2.7/site-packages/ansible/galaxy/login.py", line 100, in create_github_token
    self.remove_github_token()
  File "/home/meyers/ansible/virtualenv/ansible-dev2/local/lib/python2.7/site-packages/ansible/galaxy/login.py", line 81, in remove_github_token
    url_password=self.github_password, force_basic_auth=True,))
  File "/home/meyers/ansible/virtualenv/ansible-dev2/local/lib/python2.7/site-packages/ansible/module_utils/urls.py", line 1382, in open_url
    use_gssapi=use_gssapi, unix_socket=unix_socket, ca_path=ca_path)
  File "/home/meyers/ansible/virtualenv/ansible-dev2/local/lib/python2.7/site-packages/ansible/module_utils/urls.py", line 1235, in open
    tmp_ca_path, cadata, paths_checked = ssl_handler.get_ca_certs()
  File "/home/meyers/ansible/virtualenv/ansible-dev2/local/lib/python2.7/site-packages/ansible/module_utils/urls.py", line 870, in get_ca_certs
    to_native(b_cert, errors='surrogate_or_strict')
  File "/usr/lib/python2.7/ssl.py", line 988, in PEM_cert_to_DER_cert
    return base64.decodestring(d.encode('ASCII', 'strict'))
  File "/usr/lib/python2.7/base64.py", line 328, in decodestring
    return binascii.a2b_base64(s)
Error: Incorrect padding
```

**Comment by ansibot** [comment]
Files identified in the description:
* [`lib/ansible/galaxy`](https://github.com/ansible/ansible/blob/devel/lib/ansible/galaxy)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by jborean93** [comment]
While this might be related to certificate veritifcation I've tried to replicate this locally and I can verify that I get a proper error message. To replicate this I changed https://github.com/ansible/ansible/blob/devel/lib/ansible/galaxy/login.py#L42 to `https://<github ip>/authorizations` and then I get;

```
ansible-galaxy login -c -vvv
ansible-galaxy 2.9.0.dev0
  config file = None
  configured module search path = [u'/home/jborean/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/jborean/dev/ansible/lib/ansible
  executable location = /home/jborean/venvs/ansible-py27/bin/ansible-galaxy
  python version = 2.7.16 (default, Jul 10 2019, 13:30:25) [GCC 9.1.1 20190503 (Red Hat 9.1.1-1)]
No config file found; using defaults
Opened /home/jborean/.ansible_galaxy
ERROR! Unexpected Exception, this is probably a bug: hostname '192.30.255.116' doesn't match either of '*.github.com', 'github.com'
the full traceback was:

Traceback (most recent call last):
  File "/home/jborean/venvs/ansible-py27/bin/ansible-galaxy", line 110, in <module>
    exit_code = cli.run()
  File "/home/jborean/dev/ansible/lib/ansible/cli/galaxy.py", line 269, in run
    context.CLIARGS['func']()
  File "/home/jborean/dev/ansible/lib/ansible/cli/galaxy.py", line 888, in execute_login
    github_token = login.create_github_token()
  File "/home/jborean/dev/ansible/lib/ansible/galaxy/login.py", line 103, in create_github_token
    self.remove_github_token()
  File "/home/jborean/dev/ansible/lib/ansible/galaxy/login.py", line 84, in remove_github_token
    url_password=self.github_password, force_basic_auth=True,))
  File "/home/jborean/dev/ansible/lib/ansible/module_utils/urls.py", line 1380, in open_url
    use_gssapi=use_gssapi, unix_socket=unix_socket, ca_path=ca_path)
  File "/home/jborean/dev/ansible/lib/ansible/module_utils/urls.py", line 1286, in open
    r = urllib_request.urlopen(*urlopen_args)
  File "/usr/local/lib/python2.7/urllib2.py", line 154, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/local/lib/python2.7/urllib2.py", line 429, in open
    response = self._open(req, data)
  File "/usr/local/lib/python2.7/urllib2.py", line 447, in _open
    '_open', req)
  File "/usr/local/lib/python2.7/urllib2.py", line 407, in _call_chain
    result = func(*args)
  File "/home/jborean/dev/ansible/lib/ansible/module_utils/urls.py", line 447, in https_open
    req
  File "/usr/local/lib/python2.7/urllib2.py", line 1195, in do_open
    h.request(req.get_method(), req.get_selector(), req.data, headers)
  File "/usr/local/lib/python2.7/httplib.py", line 1042, in request
    self._send_request(method, url, body, headers)
  File "/usr/local/lib/python2.7/httplib.py", line 1082, in _send_request
    self.endheaders(body)
  File "/usr/local/lib/python2.7/httplib.py", line 1038, in endheaders
    self._send_output(message_body)
  File "/usr/local/lib/python2.7/httplib.py", line 882, in _send_output
    self.send(msg)
  File "/usr/local/lib/python2.7/httplib.py", line 844, in send
    self.connect()
  File "/home/jborean/dev/ansible/lib/ansible/module_utils/urls.py", line 429, in connect
    self.sock = self.context.wrap_socket(sock, server_hostname=server_hostname)
  File "/usr/local/lib/python2.7/ssl.py", line 369, in wrap_socket
    _context=self)
  File "/usr/local/lib/python2.7/ssl.py", line 599, in __init__
    self.do_handshake()
  File "/usr/local/lib/python2.7/ssl.py", line 836, in do_handshake
    match_hostname(self.getpeercert(), self.server_hostname)
  File "/usr/local/lib/python2.7/ssl.py", line 288, in match_hostname
    % (hostname, ', '.join(map(repr, dnsnames))))
CertificateError: hostname '192.30.255.116' doesn't match either of '*.github.com', 'github.com'
```

It looks like something is going wrong when the `open_url` is trying to read the cert chain and this error should be logged against that not `ansible-galaxy`.

In saying this, the `-c`/`--ignore-certs` option is not being passed anyway so it will always validate the cert, this just seems to be a different problem.

**Comment by sivel** [comment]
I think the following should resolve this:

```diff
diff --git a/lib/ansible/galaxy/login.py b/lib/ansible/galaxy/login.py
index dda8560991..e1b4d67746 100644
--- a/lib/ansible/galaxy/login.py
+++ b/lib/ansible/galaxy/login.py
@@ -25,6 +25,7 @@ __metaclass__ = type
 import getpass
 import json
 
+from ansible import context
 from ansible.errors import AnsibleError, AnsibleOptionsError
 from ansible.module_utils.six.moves import input
 from ansible.module_utils.six.moves.urllib.parse import quote as urlquote, urlparse
@@ -45,6 +46,7 @@ class GalaxyLogin(object):
         self.galaxy = galaxy
         self.github_username = None
         self.github_password = None
+        self._validate_certs = not context.CLIARGS['ignore_certs']
 
         if github_token is None:
             self.get_credentials()
@@ -78,7 +80,7 @@ class GalaxyLogin(object):
         '''
         try:
             tokens = json.load(open_url(self.GITHUB_AUTH, url_username=self.github_username,
-                               url_password=self.github_password, force_basic_auth=True,))
+                               url_password=self.github_password, force_basic_auth=True, validate_certs=self._validate_certs))
         except HTTPError as e:
             res = json.load(e)
             raise AnsibleError(res['message'])
@@ -88,7 +90,7 @@ class GalaxyLogin(object):
                 display.vvvvv('removing token: %s' % token['token_last_eight'])
                 try:
                     open_url('https://api.github.com/authorizations/%d' % token['id'], url_username=self.github_username,
-                             url_password=self.github_password, method='DELETE', force_basic_auth=True)
+                             url_password=self.github_password, method='DELETE', force_basic_auth=True, validate_certs=self._validate_certs)
                 except HTTPError as e:
                     res = json.load(e)
                     raise AnsibleError(res['message'])
@@ -101,7 +103,7 @@ class GalaxyLogin(object):
         args = json.dumps({"scopes": ["public_repo"], "note": "ansible-galaxy login"})
         try:
             data = json.load(open_url(self.GITHUB_AUTH, url_username=self.github_username,
-                             url_password=self.github_password, force_basic_auth=True, data=args))
+                             url_password=self.github_password, force_basic_auth=True, data=args, validate_certs=self._validate_certs))
         except HTTPError as e:
             res = json.load(e)
             raise AnsibleError(res['message'])
diff --git a/lib/ansible/module_utils/urls.py b/lib/ansible/module_utils/urls.py
index 041d0829f9..153cbc73de 100644
--- a/lib/ansible/module_utils/urls.py
+++ b/lib/ansible/module_utils/urls.py
@@ -870,7 +870,7 @@ class SSLValidationHandler(urllib_request.BaseHandler):
                                                 to_native(b_cert, errors='surrogate_or_strict')
                                             )
                                         )
-                                    except ValueError:
+                                    except Exception:
                                         continue
                                 else:
                                     os.write(tmp_fd, b_cert)
```

**Comment by jborean93** [comment]
@sivel thanks for that, it should solve `-c` not being applied to the GitHub login but I still need to figure out why the original message is failing with `Error: Incorrect padding` and not the `CertificateError` that I get locally.

**Comment by sivel** [comment]
@jborean93 when I rewrote the SSL handling, I knew there were some exceptions raised when calling `ssl.PEM_cert_to_DER_cert`.  In my testing, which was limited in scope, I only encountered a `ValueError`.  I think extending that to just `Exception` is appropriate there, as I'm not sure I can enumerate all of the exceptions that might be raised through that call stack.

It seems that for @chrismeyersfsu there must be some file that isn't what `ssl.PEM_cert_to_DER_cert` expected, but is failing in some other way, in this case with a `binascii.Error`.
